prow/io/opener.go

```diff
@@ -240,13 +240,25 @@ func (o *opener) RangeReader(ctx context.Context, path string, offset, length in
 	return reader, nil
 }
 
+var PreconditionFailedObjectAlreadyExists = fmt.Errorf("object already exists")
+
 // Writer returns a writer that overwrites the path.
 func (o *opener) Writer(ctx context.Context, p string, opts ...WriterOptions) (io.WriteCloser, error) {
+	var preconditionDoesNotExist bool
```

Comment:
I think we should do this proper by:
* changing the receiver type in the `Writer` method from `WriterOptions` to a new public inteface `type WriterOption interface { Apply(wo *WriterOptions) }`
* Unexport the existing `Apply` method on the `WriterOptions`, there is no reason for it to be exported
* Make `WriterOptions` implement the `WriterOption` interface
* start this function off by doing an `options := &WriterOptions{}; for _, opt := range opts { opt.Apply(options) }`
* Later on just do an `if options.PreconditionDoesNotExist != nil && options.PreconditionDoesNotExist`

This allows us to end up with exactly one canonical `WriterOptions` struct and thus avoids having to loop three times (1. Options that can not be directly set on the corresponding gcs/blob structs, 2. for gcs, 3. for blob)

Reply:
Interesting idea. Done